src/main/kotlin/it/pagopa/wallet/services/PaymentMethodsService.kt

```diff
@@ -0,0 +1,49 @@
+package it.pagopa.wallet.services
+
+import it.pagopa.generated.ecommerce.model.PaymentMethodResponse
+import it.pagopa.wallet.client.EcommercePaymentMethodsClient
+import it.pagopa.wallet.repositories.PaymentMethodsTemplateWrapper
+import kotlinx.coroutines.reactor.mono
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.switchIfEmpty
+
+@Service
+class PaymentMethodsService(
+    @Autowired private val paymentMethodsRedisTemplate: PaymentMethodsTemplateWrapper,
+    @Autowired private val ecommercePaymentMethodsClient: EcommercePaymentMethodsClient,
+) {
+
+    /*
+     * Logger instance
+     */
+    var logger: Logger = LoggerFactory.getLogger(this.javaClass)
+
+    fun getPaymentMethodById(paymentMethodId: String): Mono<PaymentMethodResponse> =
+        mono {
+                logger.debug("Try to retrieve payment method from cache: [$paymentMethodId]")
+                paymentMethodsRedisTemplate.findById(paymentMethodId)
+            }
+            .doOnNext { logger.info("Cache hit for payment method with id: [$paymentMethodId]") }
+            .switchIfEmpty {
+                logger.info("Cache miss for payment method: [$paymentMethodId]")
+                retrievePaymentMethodByApi(paymentMethodId)
+            }
+
+    private fun retrievePaymentMethodByApi(paymentMethodId: String): Mono<PaymentMethodResponse> =
+        ecommercePaymentMethodsClient
+            .getPaymentMethodById(paymentMethodId)
+            .doOnSuccess { savePaymentMethodIntoCache(it) }
+            .doOnError { logger.error("Error during call to payment method: [$paymentMethodId]") }
+
+    private fun savePaymentMethodIntoCache(paymentMethodResponse: PaymentMethodResponse) =
+        try {
+            logger.debug("Save payment method into cache: [${paymentMethodResponse.id}]")
+            paymentMethodsRedisTemplate.save(paymentMethodResponse)
+        } catch (e: Exception) {
+            logger.error("Error saving payment method into cache: [${paymentMethodResponse.id}]")
+        }
```

Comment:
Can we refactor this in a reactive fashion? Since paymentMethodsRedisTemplate.save is a blocking and synchronous method, we could follow the tips provided by the reactor documentation [here](https://projectreactor.io/docs/core/milestone/reference/faq.html#faq.wrap-blocking) and already applied [here](https://github.com/pagopa/pagopa-payment-wallet-scheduler-service/blob/main/src/main/kotlin/it/pagopa/wallet/scheduler/jobs/paymentwallet/OnboardedPaymentWalletJob.kt).

Reply:
Done here aaa7f564ee161da53084e5276d792dff476cd875